### PR TITLE
[ET-1541] Common Settings Image Field

### PR DESCRIPTION
**Pull request by codingmusician**
### 🎫 Ticket

[ET-1541] 

### 🗒️ Description

Adding an image field that will invoke the WP Media Gallery in order to choose/upload an image. The URL will then be stored in a hidden input that can be stored in the settings.

### 🎥 Artifacts 

https://gyazo.com/f8bfa17eb23e9ac538dae11fb90ff6fb

### ✔️ Checklist
- [x] I've included a changelog entry in the readme.txt file. 
- [x] My code is tested. 
- [x] My code has [proper inline documentation](https://the-events-calendar.github.io/products-engineering/docs/code-standards/).

[ET-1541]: https://theeventscalendar.atlassian.net/browse/ET-1541?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Tribe/Admin/Settings.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * @since TBD
+ */
+
+class Settings {
+
+    static $image_field_assets_loaded = false;
+
+    /**
+     * Loaded image field assets if not already loaded.
+     *
```
Comment (juanfra) [inline src/Tribe/Admin/Settings.php:17]: Let's add a `@since` here.

File: src/Tribe/Admin/Settings.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Admin Settings class.
+ * 
+ * @since TBD
+ */
+
+class Settings {
+
+    static $image_field_assets_loaded = false;
```
Comment (juanfra) [inline src/Tribe/Admin/Settings.php:13]: Let's document this one.

File: src/Tribe/Field.php [hunk 3]
```diff
@@ -702,6 +706,43 @@ public function license_key() {
...
+			$field .= ' type="hidden"';
+			$field .= ' class="tribe-admin-image_field-input"';
+			$field .= $this->do_field_name();
+			$field .= $this->do_field_value();
+			$field .= $this->do_field_attributes();
+			$field .= '/>';
+			$field .= '<button type="button" class="button tribe-admin-image_field-btn-add">' . $upload_image_text . '</button>';
+			$field .= '<div class="tribe-admin-image_field-image-container hidden">';
+			if ( $image_exists ) {
+				$field .= '<img src="' . $this->value . '" alt="" />';
```
Comment (juanfra) [inline src/Tribe/Field.php:735]: Do we have the alt to use here? Do we need to escape the URL just in case?

File: src/Tribe/Field.php [hunk 4]
```diff
@@ -702,6 +706,43 @@ public function license_key() {
...
+			}
+			$field .= '</div>';
+			$field .= '<a href="#" class="tribe-admin-image_field-btn-remove hidden">' . $remove_image_text . '</a>';
```
Comment (juanfra) [inline src/Tribe/Field.php:738]: Semantically this probably be a button.

File: src/resources/js/admin-image-field.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+				frame.open();
+			}
+
+			frame.off('select').on( 'select', () => {
+				var attachment = frame.state().get('selection').first().toJSON();
+				if ( imgContainer.find('img').length > 0 ){
+					imgContainer.find('img').attr('src', attachment.url);
+				}else{
+					imgContainer
+					.html( '<img src="' + attachment.url + '" alt="" />' );
```
Comment (juanfra) [inline src/resources/js/admin-image-field.js:44]: Maybe we can update the alt here as well?

File: src/resources/postcss/tribe-common-admin/_main.pcss [hunk 6]
```diff
@@ -434,6 +434,21 @@ table.plugins {
 	width: 68%;
 }
 
+.tribe-admin-image_field-btn-remove {
```
Comment (juanfra) [inline src/resources/postcss/tribe-common-admin/_main.pcss:437]: I feel the naming here should be `.tribe-admin__settings-image-field-btn-remove` (if we don't go with `tec-` as the prefix)

File: src/resources/postcss/tribe-common-admin/_main.pcss [hunk 7]
```diff
@@ -434,6 +434,21 @@ table.plugins {
...
+	color: #3c434a;
+}
+
+.tribe-admin-image_field-image-container {
```
Comment (juanfra) [inline src/resources/postcss/tribe-common-admin/_main.pcss:441]: `.tribe-admin__settings-image-field-image-container`

File: src/Tribe/Field.php [hunk 8]
```diff
@@ -702,6 +706,43 @@ public function license_key() {
 			return $field;
 		}
 
+		/**
+		 * Generate an image field.
+		 *
```
Comment (rafsuntaskin) [inline src/Tribe/Field.php:711]:
```suggestion
		 * @since TBD
```

**Review by rafsuntaskin — CHANGES_REQUESTED** [review]

File: src/Tribe/Main.php [hunk 9]
```diff
@@ -286,6 +286,23 @@ public function load_assets() {
 			]
 		);
 
+		tribe_asset(
+			$this,
+			'tribe-admin-image-field',
```
Comment (bordoni) [inline src/Tribe/Main.php:291]: lets use `tec-admin-settings-field-image`

File: src/Tribe/Main.php [hunk 10]
```diff
@@ -286,6 +286,23 @@ public function load_assets() {
...
+			[ 'jquery' ],
+			null,
+			[
+				'localize' => [
+					'name' => 'tribe_admin_image_field',
+					'data' => [
+						'select_image_text' => _x( 'Select an image to use in your email headers', 'Select image text.', 'tribe-common' ),
+						'use_image_text'    => _x( 'Use this image', 'Select image text.', 'tribe-common' ),
+					]
+				]
```
Comment (bordoni) [inline src/Tribe/Main.php:302]: Move the actual translations to the field itself as `data-` params. Allows us to have diff translations to each field.

File: src/resources/js/admin-image-field.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
+/* eslint-disable es5/no-arrow-functions */
+/* eslint-disable es5/no-es6-methods */
+jQuery( $ => {
```
Comment (bordoni) [inline src/resources/js/admin-image-field.js:3]: I love the usage of ES6, which we should continue, but we should also use the `obj` structure to expose this script globally.

Example on TEC: https://github.com/the-events-calendar/the-events-calendar/blob/master/src/resources/js/views/events-bar.js

Make sure all the callbacks are functions we can re-use.

All the selectors should be stored on `obj.selectors` so we can easily modify and indentify.

File: src/resources/js/admin-image-field.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+		addImgLink.on('click', e => {
+			e.preventDefault();
+			if ( frame ) {
+				frame.open();
+			}else{
+				// eslint-disable-next-line no-undef
+				frame = wp.media({
+					// eslint-disable-next-line no-undef
+					title: tribe_admin_image_field.select_image_text,
```
Comment (bordoni) [inline src/resources/js/admin-image-field.js:28]: Like I mentioned, pull this particular set of text variables from `data-` attrs on the main `tribe-field-image`

File: src/Tribe/Main.php [hunk 13]
```diff
@@ -286,6 +286,23 @@ public function load_assets() {
 			]
 		);
 
+		tribe_asset(
+			$this,
+			'tribe-admin-image-field',
+			'admin-image-field.js',
+			[ 'jquery' ],
+			null,
```
Comment (bordoni) [inline src/Tribe/Main.php:294]: If the conditional is being called too early, make this particular JS load later in the page like the footer o the page so you can use the style of conditional that I mentioned to you on our call.

**Review by bordoni — CHANGES_REQUESTED** [review]

File: src/Tribe/Field.php (hunk 8 above)
Comment (codingmusician) [reply src/Tribe/Field.php:713]: Good catch. Done.

File: src/Tribe/Main.php (hunk 13 above)
Comment (codingmusician) [reply src/Tribe/Main.php]: Ah.... I realized that I was trying to use `wp_footer` and I should have been using `admin_footer`. Got it fixed. Thanks for the push!

**Review by rafsuntaskin — APPROVED** [review]

File: src/Tribe/Field.php [hunk 14]
```diff
@@ -702,6 +706,52 @@ public function license_key() {
...
+
+			tribe( Settings::class )->maybe_load_image_field_assets();
+
+			$image_exists = ! empty( $this->value );
+			$upload_image_text = esc_html__( 'Select Image', 'tribe-common' );
+			$remove_image_text = esc_html__( 'Remove Image', 'tribe-common' );
+
+			// Add default fieldset attributes if none exist.
+			$image_fieldset_attributes = [
+				'data-select-image-text' => esc_html__( 'Select an image to use in your email headers', 'tribe-common' ),
```
Comment (juanfra) [inline src/Tribe/Field.php:726]: The message about the email headers should come from ET once we define the setting. For now, I'd leave a generic message to "Select an image".

File: src/resources/js/admin-image-field.js [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+	 * @since TBD
+	 * 
+	 * @param {jQuery} $fieldParent jQuery object of the field container.
+	 *
+	 * @return {void}
+	 */
+	obj.onImageSelect = function( $fieldParent ) {
+		const attachment = obj.frame.state().get('selection').first().toJSON(),
+			$imgContainer = $fieldParent.find( obj.selectors.imgContainer );
+		if ( $imgContainer.find('img').length > 0 ){
```
Comment (juanfra) [inline src/resources/js/admin-image-field.js:85]: Have we changed the JS linting rules? I see some spacing incongruencies within the JS file.
```suggestion
		if ( $imgContainer.find('img').length > 0 ) {
```

**Review by juanfra — COMMENTED** [review]
Looking good, a couple of tiny comments. I'm curious, more than anything, if we changed the js linter, as I see some spacing issues there.

File: src/resources/js/admin-image-field.js [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+	obj.addImage = function( e ) {
+		e.preventDefault();
+		const $target = $( e.target );
+		const $fieldParent = $target.closest( obj.selectors.imageFieldContainer );
+
+		if ( obj.frame ) {
+			obj.frame.open();
+		} else {
+			obj.frame = wp.media({ // eslint-disable-line no-undef
+				title: $fieldParent.data( 'select-image-text'),
```
Comment (juanfra) [inline src/resources/js/admin-image-field.js:112]:
```suggestion
				title: $fieldParent.data( 'select-image-text' ),
```

File: src/resources/js/admin-image-field.js [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+				button: {
+					text: $fieldParent.data( 'use-image-text')
```
Comment (juanfra) [inline src/resources/js/admin-image-field.js:114]:
```suggestion
					text: $fieldParent.data( 'use-image-text' )
```

File: src/resources/js/admin-image-field.js [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+				},
+				multiple: false
+			});
+			obj.frame.open();
+		}
+
+		obj.frame.off('select').on( 'select', function() {
```
Comment (juanfra) [inline src/resources/js/admin-image-field.js:121]:
```suggestion
		obj.frame.off( 'select' ).on( 'select', function() {
```

File: src/resources/js/admin-image-field.js [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+	 * @since TBD
+	 * 
+	 * @param {jQuery} $fieldParent jQuery object of the field container.
+	 *
+	 * @return {void}
+	 */
+	obj.onImageSelect = function( $fieldParent ) {
+		const attachment = obj.frame.state().get('selection').first().toJSON(),
+			$imgContainer = $fieldParent.find( obj.selectors.imgContainer );
+		if ( $imgContainer.find('img').length > 0 ) {
```
Comment (juanfra) [inline src/resources/js/admin-image-field.js:85]:
```suggestion
		if ( $imgContainer.find( 'img' ).length > 0 ) {
```

**Review by juanfra — APPROVED** [review]

**Review by juanfra — APPROVED** [review]

**Pull request merged**